lumicks/pylake/detail/tests/test_value.py

```diff
@@ -0,0 +1,183 @@
+    assert a <= 5
+    assert 5 <= a
+
+
+def test_operators_str():
+    a = Parameter(a0 := "a", "alpha")
+    b = Parameter(b0 := "b", "beta")
+
+    assert a0 + "c" == a + "c"
+    assert "c" + a0 == "c" + a
+    assert a0 + b0 == a + b
+    assert a0 * 5 == a * 5
+    assert a0 * 5 == a * Parameter(5, "five")
+    assert a0 * Parameter(5, "five") == a * 5
+
+    assert a == "a"
+
+    with pytest.raises(TypeError, match="not all arguments converted during string formatting"):
+        a % 1
+
+    assert "Hello, venus" == Parameter("Hello, %s", "fmt") % "venus"
+
+    for c in (1, "c", True):
+        for op in operators - {operator.add, operator.mul, operator.mod}:
+            with pytest.raises(TypeError):
+                op(a, c)
+
+
+def test_operators_bool():
+    a = Parameter(a0 := True, "true")
+
+    for op in operators:
+        assert op(a0, True) == op(a, True)
+        assert op(False, a0) == op(False, a)
+
+
+def test_not_implemented_operators():
+    a = Parameter(a0 := 2, "two")
+
+    for op in (operator.lshift, operator.rshift):
+        with pytest.raises(TypeError):
+            op(a, 2)
+        with pytest.raises(TypeError):
+            op(2, a)
+
+    # __matmul__ has not been specifically implemented, but it still works
+    A0 = np.array([[1, a0], [3, 4]])
+    A = np.array([[1, a], [3, 4]])
+    B0 = np.array([[11, 12], [13, 14]])
+    np.testing.assert_equal(A @ B0, A0 @ B0)
+
+    with pytest.raises(ValueError, match="matmul: Input operand 1 does not have enough dimensions"):
+        A @ 2
+
+    with pytest.raises(ValueError, match="matmul: Input operand 0 does not have enough dimensions"):
+        2 @ A
+
+    # not directly implemented but default fallback means it still works
+    assert isinstance(a, Parameter)
+    a0 += 4
+    a += 4
+    assert a0 == a
+    assert isinstance(a, int)
+
+
+def test_casting():
+    a = Parameter(a0 := 2, "two")
+    assert float(a0) == float(a)
+
+    b = Parameter(b0 := 3.14, "pi")
+    assert int(b0) == int(b)
+
+    assert complex(a0) == complex(a)
+    assert complex(b0) == complex(b)
+    assert complex(a0, 2.1) == complex(a, 2.1)
+    assert complex(b0, 2.1) == complex(b, 2.1)
+
+
+def test_other():
+    a = Parameter(a0 := 2, "two")
+    assert -a0 == -a
+
+    b = Parameter(b0 := -2, "two")
+    assert +b0 == +b
+
+    assert abs(b0) == abs(b)
+
+    p = Parameter(p0 := 3.141592, "pi")
+    assert round(p0) == round(p)
+    assert round(p0, 3) == round(p, 3)
+
+    assert math.trunc(p0) == math.trunc(p)
+    assert math.floor(p0) == math.floor(p)
+    assert math.ceil(p0) == math.ceil(p)
+
+
+def test_numpy():
+    a = Parameter(a0 := 5, "five")
+    arr = np.arange(3) + 1
+
+    with pytest.raises(TypeError):
+        divmod(arr, a)
+
+    for op in operators - {divmod}:
+        np.testing.assert_equal(op(arr, a), op(arr, a0))
+        np.testing.assert_equal(op(a, arr), op(a0, arr))
+
+    arr = np.array([1, 2, a])
+    assert isinstance(arr[-1], Parameter)
```

Comment:
Is this type preservation inside a numpy array something you specifically want?

Because if not, it might be worth just implementing `__array__` as:
```python
    def __array__(self):
        return np.asarray(self.value)
```
Which would make things like `np.testing.assert_allclose()` and various other numpy methods work.